File: Delphi/uMainForm.pas
unit uMainForm;

interface

uses
  Winapi.Windows, Winapi.Messages, System.SysUtils, System.Variants, System.Classes, Vcl.Graphics,
  Vcl.Controls, Vcl.Forms, Vcl.Dialogs, cxClasses, dxReport, cxGraphics,
  cxLookAndFeels, cxLookAndFeelPainters, Vcl.Menus,  cxButtons,
  dxReport.Backend, cxControls, cxStyles, cxCustomData, cxFilter, cxData,
  cxDataStorage, cxEdit, cxNavigator, dxDateRanges, dxScrollbarAnnotations,
  Data.DB, cxDBData, cxGridLevel, cxGridCustomView, cxGridCustomTableView,
  cxGridTableView, cxGridDBTableView, cxGrid, FireDAC.Comp.DataSet,
  FireDAC.Comp.Client, dxReport.ConnectionString.JSON.DB,
  dxReport.ConnectionString.JSON, Vcl.StdCtrls, dxmdaset, dxShellDialogs,
  dxReport.ConnectionString.SQL, dxLayoutControlAdapters, dxLayoutContainer,
  dxLayoutControl, dxCore, dxSkinsForm, System.ImageList, Vcl.ImgList,
  cxImageList, 
  dxLayoutcxEditAdapters, cxContainer, cxTextEdit;

type
  TMainForm = class(TForm)
    dxReport1: TdxReport;
    btnPreview: TcxButton;
    dxReportDataConnectionManager: TdxReportDataConnectionManager;
    ReportsNWindConnectionString: TdxReportDatabaseSQLConnection;
    dxLayoutControl1Group_Root: TdxLayoutGroup;
    dxLayoutControl1: TdxLayoutControl;
    liPreview: TdxLayoutItem;
    dxSkinController1: TdxSkinController;
    dxLayoutLabeledItem1: TdxLayoutLabeledItem;
    dxLayoutGroup1: TdxLayoutGroup;
    dxLayoutGroup2: TdxLayoutGroup;
    gvProducts: TcxGridDBTableView;
    cxGrid1Level1: TcxGridLevel;
    cxGrid1: TcxGrid;
    dxLayoutItem1: TdxLayoutItem;
    dxLayoutGroup3: TdxLayoutGroup;
    DataSource1: TDataSource;
    gvProductsProductID: TcxGridDBColumn;
    gvProductsProductName: TcxGridDBColumn;
    gvProductsQuantityPerUnit: TcxGridDBColumn;
    gvProductsUnitsInStock: TcxGridDBColumn;
    gvProductsCategoryID: TcxGridDBColumn;
    gvProductsCategoryName: TcxGridDBColumn;
    mdProducts: TdxMemData;
    mdProductsProductID: TIntegerField;
    mdProductsProductName: TWideStringField;
    mdProductsQuantityPerUnit: TWideStringField;
    mdProductsUnitsInStock: TSmallintField;
    mdProductsCategoryID: TIntegerField;
    mdProductsCategoryName: TWideStringField;
    NWindMemDataConnection: TdxReportDataSetJSONConnection;
    NWindMemDataConnectionItem1: TdxReportDataSetCollectionItem;
    procedure btnPreviewClick(Sender: TObject);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  MainForm: TMainForm;

implementation

{$R *.dfm}

procedure TMainForm.btnPreviewClick(Sender: TObject);
var
  i: Integer;
  ProductIDs: TStringList;
  RowIndex: Integer;
begin
  ProductIDs := TStringList.Create;
  try
    for i := 0 to gvProducts.Controller.SelectedRowCount - 1 do
    begin
      RowIndex := gvProducts.Controller.SelectedRows[i].RecordIndex;
      ProductIDs.Add(gvProducts.DataController.Values[RowIndex, gvProductsProductID.Index]);
    end;

    if ProductIDs.Count > 0 then
      dxReport1.FilterString := Format('[ProductID] In (%s)', [ProductIDs.CommaText])
    else
      dxReport1.FilterString := '';

    dxReport1.ShowViewer;
  finally
    ProductIDs.Free;
  end;
end;

end.

File: CPB/uMainForm.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "uMainForm.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxButtons"
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDataStorage"
#pragma link "cxDBData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxGrid"
#pragma link "cxGridCustomTableView"
#pragma link "cxGridCustomView"
#pragma link "cxGridDBTableView"
#pragma link "cxGridLevel"
#pragma link "cxGridTableView"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxLookAndFeels"
#pragma link "cxNavigator"
#pragma link "cxStyles"
#pragma link "dxDateRanges"
#pragma link "dxReport"
#pragma link "dxReport.Backend"
#pragma link "dxReport.ConnectionString.JSON"

#if defined(_WIN64)
  #pragma link "dxReport.ConnectionString.JSON.DB.O"
#else
  #pragma link "dxReport.ConnectionString.JSON.DB.OBJ"
#endif


#pragma link "dxScrollbarAnnotations"
#pragma link "dxmdaset"
#pragma link "dxShellDialogs"

#pragma link "dxCore"
#pragma link "dxLayoutContainer"
#pragma link "dxLayoutControl"
#pragma link "dxLayoutControlAdapters"
#pragma link "dxSkinsForm"
#pragma resource "*.dfm"
TMainForm *MainForm;
//---------------------------------------------------------------------------
__fastcall TMainForm::TMainForm(TComponent* Owner)
	: TForm(Owner)
{

}
//-----------------------------------------------------------------

void __fastcall TMainForm::btnPreviewClick(TObject *Sender)
{
	TStringList *ProductIDs = new TStringList();
	try
	{
		for (int i = 0; i < gvProducts->Controller->SelectedRowCount; i++)
		{
			int RowIndex = gvProducts->Controller->SelectedRows[i]->RecordIndex;
			ProductIDs->Add(
				gvProducts->DataController->Values[RowIndex][gvProductsProductID->Index]
			);
		}

		if (ProductIDs->Count > 0)
			dxReport1->FilterString = AnsiString::Format("[ProductID] In (%s)", ARRAYOFCONST((ProductIDs->CommaText)));
		else
			dxReport1->FilterString = "";

		dxReport1->ShowViewer();
	}
	__finally
	{
		delete ProductIDs;
	}
}



//---------------------------------------------------------------------------

